A roadmap for package infrastructure without astropy-helpers

astrofrog opened the pull request:
This is an APE describing a roadmap for transitioning away from astropy-helpers in favor of more standardize tools in the Python ecosystem. The short version is that:

* Pure-Python packages that use astropy-helpers currently can already be refactored to not rely on astropy-helpers, and in fact several packages have already done so (see links in the implementation section).

* Packages with C extensions can start to make some of the changes here but should wait a little before making the more significant changes related in particular to build-time dependencies since they require a recent version of pip (19.0 or later) - but we wanted to start the discussion now to start planning the changes.

A fraction of astropy-helpers would continue to exist as a more general extension-helpers package specifically geared towards packages with compiled extensions, and this could be used outside the astropy project, and would no longer be included as a sub-module.

Even though we can't make some of these changes in the core package yet, we wanted to start the discussion now so as to be able to make sure there are no issues in the current Python packaging ecosystem that would prevent these changes.

This is just a draft at this point, and we welcome detailed comments! If some points prove to be controversial, we'll try and set up a call so that we can discuss things and see if we can reach a consensus! Please also say if any aspects of this is confusing - some of us have spent far too much time thinking about this, and there may be some points that could be explained better.

bsipocz commented on APE17.rst:
This docs washer over on the fact that the current `python setup.py test` solution provides access to all the conveniences the astropy test runner has. Neither tox nor pytest provides any of those, so this proposal means removal of features many of the core devs relied on (e.g. don't underestimate the importance of e.g. `-P table,units` picks up all the tests from both subpackages and their documentations. It's not just much more typing but more error prone to need to think about where all the relevant tests live in a repo).

astrofrog replied:
As mentioned above, we could add this back as a pytest option. What other test runner options do you find useful?

bsipocz replied:
`-R` is another such nice conveniences

bwinkel replied:
What about the `--remote-data` option? Also, when I develop tests, I like to use the pytest options, which can be used via `python setup.py test -a <pytest args>`, e.g., `-x` to exit after first failure. (Several of my packages have C extensions, so running pytest directly wouldn't be possible.)

bsipocz replied:
`-R` is short for `--remote-data` 😉

bwinkel replied:
I know, it was just a race condition 😁

astrofrog replied:
Ok I've added a section on this, and a link to my PR to pytest-astropy to add the ``-P`` option for pytest: https://github.com/astropy/pytest-astropy/pull/19

astrofrog replied:
I've added a table showing that most of the previous options/conveniences translate over to options with pytest: https://github.com/astropy/astropy-APEs/pull/52/commits/8841b8aea62edab63cf19e029c96922cb1915f55

astrofrog commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+However, for pure-Python packages, developers can also opt to simply run the
+tests directly with pytest, e.g.::
+
+    pytest packagename
```
We could always add these back via a pytest plugin in astropy (like the display plugin at the moment which all affiliates have access to)

bsipocz replied:
for the record, `-R` is another such nice conveniences

astrofrog replied:
I've now added a table of correspondence for ``setup.py test`` to ``pytest`` options in https://github.com/astropy/astropy-APEs/pull/52/commits/8841b8aea62edab63cf19e029c96922cb1915f55 - and mention my PR for the ``-P`` option.

mhvk replied:
Thanks for the table! I agree with @bsipocz that the `-P` option is the one that I feel I would miss most, so :+1: to having that as part of the `astropy-pytest` (or so) plugin.

astrofrog commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+a clean environment, it ensures that source distributions for the package are
+fully functional and that all required files and dependencies are declared. For
+packages with C/Cython extensions it should also guarantee that they are always
+rebuilt before running the tests.
```
This was done before with ``python setup.py test`` too by the way, it's not something that would be new with tox. But I always run tests by doing:

```
python setup.py build_ext --inplace  # just do this once
pytest astropy/table
```

for example. Then unless I change the C code I can just use pytest directly on the source in the repo.

bsipocz replied:
so yo don't test the narrative docs?

bsipocz replied:
> python setup.py build_ext --inplace  # just do this once
pytest astropy/table

a way to automate this would be great (e.g. the behaviour we have atm).

astrofrog replied:
I now mention the ``pytest-build`` plugin to do in-place builds

astrofrog replied:
I've also changed the ``pytest`` command in the APE from ``pytest package`` to just ``pytest`` which will allow the rst docs to get picked up too.

astrofrog commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+The original APE proposing the creation of astropy-helpers (`APE 4`_) was
+accepted in June 2014. Since then, the state of the Python packaging ecosystem
+has changed significantly, with advances such as binary wheels that can be
+provided for all major platforms, tools such as `tox`_ and `setuptools_scm`_,
```
Good point, we can rephrase this.

astrofrog replied:
Actually setuptools_scm was only released in 2015, but tox has been around a lot longer

taldcroft commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+
+The benefit of using tox over just running pytest directly is that since behind
+the scenes it creates a source distribution of the package and installs it into
+a clean environment, it ensures that source distributions for the package are
```
Does this mean it will need to rebuild the entire astropy every time you want to run tests in this way?

taldcroft replied:
Sorry, should have scrolled down two lines further.  So I think the answer is that developers need to manage the build themselves in order to do repetitive light-weight testing.

astrofrog replied:
I've now added some notes about this and what solutions exist to make this as easy as possible.

mwcraig commented on APE17.rst:
Looks like you haven't defined the reference for this pip link.

astrofrog replied:
Fixed

mwcraig commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+such as the core astropy package which have all their tags on branches (at least
+in recent years), we will need to add a ‘developer’ tag on ``master`` straight
+after branching, e.g. after creating a ``v4.0.x`` branch we should tag the next
+commit on ``master`` as ``v4.1.dev``.
```
It seems like this should be noted somewhere, though, or it will be forgotten in the future. Maybe part of release instructions?

eteq replied:
It also seems a bit problematic because it means *affiliated* packages would have to start doing this.  And many affiliated packages do not have standardized release procedures/instructions...

astrofrog replied:
No I don't think affiliated packages would need to do this. This issue only occurs for packages that have a workflow where they tag releases in branches, but in my experience most affiliated packages don't do release branches.

astrofrog replied:
I think we should leave this in the APE - the APE is *not* documentation but just a technical document for the project, and I think it's important to note this here. But I've added a sentence that says

*In any case, the documentation about how to release the core package as well astropy-affiliated packages will need to be updated to reflect the use of setuptools_scm.*

astrofrog replied:
We are including instructions about how to use setuptools_scm with release branches in an upcoming packaging guide.

olebole commented:
+1 since it simplifies packaging.

hamogu commented:
+1 I also note that of the 400 packages that use astropy-helpers, I'm probably responsible for 8, only 2 of which are active and one 1 of which is not pure-python. (I'll try to remember to delete experiemental remaining packages from github ...)

mhvk commented:
Another :+1: though like @bsipocz I use `./setup.py test -P <module>` a *lot* (also for my own package), and like that it does the docs as well; I also use `--remote-data` somewhat regularly.

dpshelio commented on APE17.rst:
:point_up_2: @giordano @helgee @kbarbary are you really planning to take over?

giordano replied:
:joy:
![](https://i.imgur.com/KgEOkS1.gif)

pllim commented on APE17.rst:
I am not convinced yet another package is necessary. Instead, we can show them how things should be done "natively" over at template?

eteq replied:
I don't quite understand how this would work as @pllim is proposing.  Then wouldn't we have to push updates to *all* users of this stuff if e.g. we need to update the `OpenMP` stuff?

astrofrog replied:
Our draft packaging guide shows both how to define extensions natively for simple cases and with extension-helpers:

https://oa-packaging-guide-preview.readthedocs.io/en/latest/extensions.html

For astropy, it's definitely the case that we have enough extensions that using extension-helpers would be a lot simpler than defining everything in the setup.py file, and the auto-detection of cython files is also a nice feature.

@pllim - are you satisfied with this or do you have specific wording recommendations for the APE?

Cadair replied:
I think for packages that have one or two C extensions might not need to use `extension_helpers` as the standard ways of registering extensions would be sufficient. Certainly for core I think `extensions_helpers` would simplify things enough to warrant it's existence.

pllim replied:
Okay. I think I understand now. Thanks!

pllim commented on APE17.rst:
```suggestion
Therefore, we recommend that pure-Python packages can make all the changes
```

astrofrog replied:
Fixed

pllim commented on APE17.rst:
```suggestion
source distributions, but they can still make other changes now, e.g., using
```

astrofrog replied:
Fixed

pllim commented on APE17.rst:
```suggestion
``python setup.py test`` or ``python setup.py build_docs`` gives a helpful
```

astrofrog replied:
Fixed

StanczakDominik commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+
+developers will then easily be able to run the tests with::
+
+    tox -e test
```
I recently restructured PlasmaPy's CI setup to use Tox following SunPy's example - there was nothing I needed to change in the tests themselves.

astrofrog replied:
@StanczakDominik is right, this doesn't require restructuring the tests - all we need is to write a ``tox.ini`` file but no changes are needed to the package or tests.

astrofrog commented on APE17.rst:
There have been some issues with some versions of pip, but I don't think these would affect a simple case like what we need here.

astrofrog commented on APE17.rst:
Fixed

astrofrog commented on APE17.rst:
```diff
@@ -0,0 +1,482 @@
+pytest-astropy if needed, and ran the tests there.
+
+However, this can now be done more cleanly using `tox`_. Tox is a tool that
+makes is easy to install packages into a temporary clean Python environment
```
Fixed

astrofrog commented on APE17.rst:
it shouldn't be used at all, I've clarified

astrofrog commented:
I've made the following significant additions based on feedback above:

* I've added a recommendation that ``python setup.py test`` and ``python setup.py build_docs`` should not just crash but instead give useful information on how to run the tests and build the docs, and also provide a link to more detailed documentation online.

* I've added a mention of how developers could use pytest (for both pure-python packages and packages with extensions) without having to use tox. I've still made sure tox is the primary recommended way, but acknowledging that developers sometimes want to just do quick development/testing with pytest directly.

* I've added a mention of how we can replicate options like ``-P`` and ``-R`` with pytest and tox the latter of which can just pass command-line arguments to pytest).

* I've mentioned more explicitly that the release docs will need to be updated to reflect the updated workflow with setuptools_scm

* I've also made small updates to wording in places based on feedback here.

karllark commented:
Easier and more generic package maintenance is something I support.  The overhead of keeping up with this on my half dozen packages that use the package template/astropy-helpers is increasing with time.   Thumbs up!

bsipocz commented:
> Easier and more generic package maintenance is something I support. The overhead of keeping up with this on my half dozen packages that use the package template/astropy-helpers is increasing with time. Thumbs up!

I don't think it will get easier to keep generic files in sync with the template than using a submodule that is versioned and understood not to be edited (auto updating the helpers is always way easier than any of the other auto updates we do, as folks might have edited those files, but not the submodule). 
But either way, getting rid of the submodule, things definitely become more intuitive as the concept of submodules are not that widely known.

And of course it gets easier for us maintainers to not support functionality that is now widely available in generic tools.

pllim commented:
Can we accept this already? :smile: 

p.s. When it is time... https://www.atlassian.com/blog/git/git-submodules-workflows-tips#tip2

bsipocz commented:
@pllim - first we said that the PRs should be set up and work for a couple packages.

pllim commented:
Re: the PRs -- Ah okay, I see @larrybradley is already working on `photutils`. astropy/photutils#915

bsipocz commented:
yes, @astrofrog mentioned one being planned for ccdproc, too, and I might be overly ambitious, but having seen one for astropy would be nice (in addition to https://github.com/astropy/pytest-astropy/pull/19 being merged).

pllim commented:
So, @astrofrog , can you please list the PRs as acceptance criteria check list on the original post, and then when they are all checked off, we accept this?

astrofrog commented:
Just to avoid duplicated effort, I do have a branch of removing astropy-helpers in the core package here: https://github.com/astrofrog/astropy/pull/83 - but needs cleaning up.

Ultimately the CoCo will need to make the final decision, but I'll send another email out to astropy-dev to let people know I've updated it and it's ready for a re-review for last objections.

cdeil commented:
@astrofrog - Thanks!

I'm removing `astropy-helpers` from Gammapy: https://github.com/gammapy/gammapy/pull/2279

I tried adding `pyproject.toml`, but then what happened (I think) was that `pip install -e .` was creating a virtualenv with Numpy and Cython and using that to run the build. This is not what I want, I find this annoying. We have a [environment-dev.yml](https://github.com/gammapy/gammapy/blob/master/environment-dev.yml) and devs use that, it has Numpy and Cython. Is there a way to have a `pyproject.toml` but still by default use the existing Python env for developers? For released versions, it would be nice to have a `pyproject.toml`, and to have `pip install gammapy` work, instead of getting a Numpy or Cython related import error.

Another point is that we have a [Makefile](https://github.com/gammapy/gammapy/blob/master/Makefile) and just use the default dev env locally for tests, docs, etc. Some day we might start using tox, but for now what we have is fine and none of our devs knows or likes tox. One thing I'd suggest for the APE is to make it a bit clearer that e.g. using tox is a recommendation, not a requirement, to get rid of astropy-helpers.

One more thing that's unclear to me: what about in-progress pull requests and developer git checkouts? What commands do they have to run to get rid of the astropy-helpers git submodule? I think they have to manuallly remove it, and then if they have a mix of local branches before or after the `astropy-helpers` removal, they are in trouble? Maybe add a few lines and commands they need during the transition period (until there are no more branches with `astropy-helpers` around for a given project)?

astrofrog commented:
@cdeil - thanks for trying this out!

> I tried adding pyproject.toml, but then what happened (I think) was that pip install -e . was creating a virtualenv with Numpy and Cython and using that to run the build. This is not what I want, I find this annoying. We have a environment-dev.yml and devs use that, it has Numpy and Cython.

The isolated build environment is a feature recommended by PEP517 - see [here](https://www.python.org/dev/peps/pep-0517/#recommendations-for-build-frontends-non-normative), specifically:

*A build frontend SHOULD, by default, create an isolated environment for each build, containing only the standard library and any explicitly requested build-dependencies.*

However I thought that when doing ``-e`` pip didn't enforce build isolation - maybe @Cadair has some thoughts on this?

> Is there a way to have a pyproject.toml but still by default use the existing Python env for developers? For released versions, it would be nice to have a pyproject.toml, and to have pip install gammapy work, instead of getting a Numpy or Cython related import error.

It sounds like something weird is going on here - normally your pyproject file should declare numpy and cython as build dependencies, and your package should also define numpy as an install_requires, and you shouldn't get any errors related to numpy or cython. Can you point me to the error you were seeing?

> Another point is that we have a Makefile and just use the default dev env locally for tests, docs, etc. Some day we might start using tox, but for now what we have is fine and none of our devs knows or likes tox. One thing I'd suggest for the APE is to make it a bit clearer that e.g. using tox is a recommendation, not a requirement, to get rid of astropy-helpers.

Sounds good, I'll try and clarify that it's recommended but optional.

> One more thing that's unclear to me: what about in-progress pull requests and developer git checkouts? What commands do they have to run to get rid of the astropy-helpers git submodule? I think they have to manuallly remove it, and then if they have a mix of local branches before or after the astropy-helpers removal, they are in trouble? Maybe add a few lines and commands they need during the transition period (until there are no more branches with astropy-helpers around for a given project)?

Sounds good, submodules are indeed a pain to remove so I'll provide some clear instructions.

astrofrog commented:
Just as a reminder to myself, other projects such as Numpy are now seeing issues with shipping pre-cythonized files (https://github.com/numpy/numpy/issues/14403) which the approach recommended in the APE should fix.

cdeil commented:
It turned out that there were no transition issues with removing the git sub-package. All devs had to do was `rm -r astropy_helpers` and even switching back and forth with branches that had or didn't have the sub-module any more just worked (see 
 https://github.com/gammapy/gammapy/pull/2279#issuecomment-525180658).

We had one more small issue where we didn't manage to run `sphinx-build` from the top-level project folder to get rid of a sphinx warning, we had to `cd docs` and run the sphinx build there (see https://github.com/gammapy/gammapy/pull/2279#issuecomment-512762075). This could be worth mentioning somewhere (in the project template)?

Concerning `pyproject.toml` I looked around a bit:
- It seems that in April & May 2019 the main issues with `pip install -e` and `pyproject.toml` were resolved (see [here](At https://discuss.python.org/t/pip-19-1-and-installing-in-editable-mode-with-pyproject-toml/1553)).
- pip now has options [--no-use-pep517](https://pip.pypa.io/en/stable/reference/pip_install/#cmdoption-use-pep517) and [--no-build-isolation](https://pip.pypa.io/en/stable/reference/pip_install/#cmdoption-no-build-isolation)

So it looks like the solution for a project like Gammapy where a conda dev env is provided, if we want to add `pyproject.toml`, would be to recommend the use of `pip install --no-build-isolation -e .` (or put that command as `make install-dev` in our Makefile to make it slightly easier to remember and type).

I still don't understand why we'd want to pin `numpy==1.14.2` for the build. Why should a user have to download and build an old Numpy when they `pip install gammapy`, wouldn't it be better to use the same Numpy that they have in the env where Numpy & Astropy & Gammapy is being installed. Why can't pip simply use the info and install Numpy first? Is it really the idea to always install two Numpy on all user machines just to install a package that has Cython & Numpy code?

astrofrog commented:
> I still don't understand why we'd want to pin numpy==1.14.2 for the build

Just to be clear, this version is *only* used for building the package, and is not installed into the user's environment. The idea is to build the package with the oldest compatible version of Numpy. So for example, if you were to just put 'numpy' (with no version pinning) in pyproject.toml and 'numpy' in install_requires, then you could run into the following if you started with a clean environment

* ``pip install gammapy`` would temporarily download Numpy 1.17, build a gammapy wheel with that version of Numpy, then install the wheel into the user environment along with Numpy 1.17. So far so good.
* Now the user decides they want to downgrade Numpy to 1.15 because of some of the code, so ``pip install "numpy==1.15.*"``. Now if they try and import gammapy C extensions, they will get numpy errors since there was a compatibility break in 1.16.

So by pinning the oldest compatible version in ``pyproject.toml``, you effectively do the following:

* ``pip install gammapy`` would temporarily download Numpy 1.14, build a gammapy wheel with that version of Numpy, then install the wheel into the user environment along with Numpy **1.17** (taken from install_requires).
* Now the user decides they want to downgrade Numpy to 1.15 because of some of the code, so ``pip install "numpy==1.15.*"``. Now if they try and import gammapy C extensions, everything will work fine. In fact, they can downgrade Numpy to any compatible version and everything should be ok.

Does this make sense? The key here is that the version in ``pyproject.toml`` is just used when building a wheel, and in the same way that when building wheels for releases we need to use the oldest compatible Numpy version, we need to do this here too.

cdeil commented:
@astrofrog - So the Numpy version pinning makes sense when building and distributing binary wheels.

That's not so common for devs, I'm not sure using `pyproject.toml` for that use case is a great idea.

If you want to build and distribute binary wheels, then you probably want to do it for different platforms, and you don't want to do it on a dev machine. Rather you want to have something like the `meta.yml` in conda-forge, and their tooling. That's where the Numpy version is specified to be compatible across conda-forge packages, it's not using `pyproject.toml`.

Now pip also builds binary wheels locally and installs those. For that use case, I prefer if the Numpy from my existing environment is used. This way I avoid old and possibly buggy or not as performant Numpy versions (not sure if that matters in practice), and just the extra time spent to make a separate build env and to install Numpy.

Note that it's not obvious which Numpy to pin against. E.g. in this APE you have `numpy=1.13.3` in the example. I'm on Python 3.7, and for that Numpy no binary wheels were build: https://pypi.org/project/numpy/1.13.3/#files
Now for me on macOS today, `pip` did manage to build that old Numpy from source and I think it linked against the Apple BLAS libs which is no longer recommended. But I can imagine there's a pretty good chance this install of old Numpy from source will fail on some machines.

While I love the idea of having a `pyproject.toml` to solve the problem we had forever in Python that `pip` didn't install Numpy before installing packages with C extensions, I'm not sold on the idea of this separate build env and doing extra builds with old Numpy versions on user machines.

astrofrog commented:
@cdeil - I understand your concerns, so I'll be sure to mention the flag to skip the build isolation.

> Note that it's not obvious which Numpy to pin against. E.g. in this APE you have numpy=1.13.3 in the example

Search for PEP 508 in this APE, as this will show an example of how to specify different versions of Numpy versions for different Python versions using environment markers. The oldest version that should be specified should be the oldest one with wheels for that Python version.

> While I love the idea of having a pyproject.toml to solve the problem we had forever in Python that pip didn't install Numpy before installing packages with C extensions, I'm not sold on the idea of this separate build env and doing extra builds with old Numpy versions on user machines.

One interesting point is that since most *users* actually install from wheels (and therefore don't have to do builds anyway) most users will be using packages built with the oldest compatible Numpy version and with a different installed Numpy version, so in fact if you use the oldest Numpy in pyproject.toml you are actually doing something that will be closer to the average user of your package.

Cadair commented:
> This way I avoid old and possibly buggy or not as performant Numpy versions (not sure if that matters in practice)

I think where this matters would be a pretty small subset of changes to Numpy. What you are doing by specifying an old version of numpy is building your C extensions against the oldest version of the Numpy ABI. The actual code you execute will be the one in your final installed environment. So the only case you would take a performance penalty that I can think of is the case where worse performing code is kept in Numpy for the sole reason of ABI compatibility.

dhomeier commented:
> Just as a reminder to myself, other projects such as Numpy are now seeing issues with shipping pre-cythonized files ([numpy/numpy#14403](https://github.com/numpy/numpy/issues/14403)) which the approach recommended in the APE should fix.

Cf. my message on slack - looks like the numpy issue was just due to running Cython 0.29.11 instead of 0.29.13 on them; so hopefully it will still be possible to create sources that work with Pythons 3.6-3.8+, if necessary.

astrofrog commented:
@cdeil - I've tried to clarify some of the points you raised - you can see just the changes relevant to your comments at https://github.com/astropy/astropy-APEs/pull/52/commits/ff6d75fea1fcb3480f5fa082f8b06bc8ceaa6abe?short_path=7e96e0f#diff-7e96e0f54f0335dc6dfa73709a7c40d7

cdeil commented:
@astrofrog - Thanks!

mhvk commented:
I have been following this on and off, as I'd definitely love not to have a submodule anymore, but am also more than a little used to the current workflow. Reading the whole APE again, I see that much of my initial worries have now been addressed -- I'm almost ready to get rid of `astropy-helpers` in my own packages! 

But the on thing that is holding me back is still what I use most -- `./setup.py test`. Like @cdeil, I think it is important to distinguish testing during development and testing for continuous integration. In particular, during development, one really does not want to be slowed down by repeatedly installing in a clean environment, so it would seem that for that case using `pytest` directly is what should be recommended. Especially for astropy, one also does *not* want to test the whole package, but rather the submodule one is working on. Hence, preserving the ability to test a submodule with `-P` or equivalent is really important, and it would be very helpful if it continues to be possible to test the associated documentation at the same time. (For complete tests done before submitting PRs and for continuous integration, `tox` seems to be a good replacement.)

Also, a question: since it is already mentioned that one should let the old `setup.py` commands emit text that tells about the new state, and that, therefore, it is necessary to make a more complicated `setup.py`, would it not be possible instead to just let those commands *use* the new recommended procedure? I.e., could one get away with replacing `astropy-helpers` with a more complicated `setup.py`? I realize this partially beats the whole point, but if it means that for most users the transition is essentially invisible, that would be worth quite a bit too.

astrofrog commented:
@mhvk - could you try out the following plugin which adds a ``-P`` option for pytest: https://github.com/astropy/pytest-astropy/pull/19

Also I have tried to make it clear in the current wording that developers can choose to just run pytest directly:

```
For developers who do not want to always use tox to run the test, it is also
possible to run pytest directly. For pure-Python packages, developers can run,
e.g.::

    pytest packagename

For developers working on packages with compiled extensions, the preferred
solution is to install the package in editable mode before running the tests::

    pip install -e .
    pytest packagename

Alternatively, one can build the extensions in-place before running the tests::

    python setup.py build_ext --inplace
    pytest packagename

The latter can also be simplified using a plugin such as `pytest-build
<https://github.com/astrofrog/pytest-build>`_ which allows this example to be
written as::

    pytest packagename --build-inplace
```

and in fact this will be faster than the current ``python setup.py test`` since that was installing into a temporary directory every time. If you think we could phrase this better though, could you open a PR with suggestions for how to word things differently? For any packages other than the core package, tox will be fast though so I don't see the downside in still recommending that - there's a misconception that it's a CI tool when in fact it's just as useful for local use and it catches a number of packaging bugs that otherwise might go unnoticed (I've found issues on many packages where I've added a tox config).

For continuing to support ``python setup.py test`` as an alias, the issue is that we can't just rely on just wrapping pytest since that doesn't set up things like entry points directly, so we would then need to keep our whole test runner infrastructure which installs to a temporary directory (and one of the points of this APE was to reduce the amount of custom infrastructure we have to support) - or we could wrap tox, but this is explicitly discouraged [in the tox docs](https://tox.readthedocs.io/en/latest/example/basic.html#integration-with-setup-py-test-command). In addition, the idea is that in future ``setup.py`` commands are going to become discouraged anyway in Python packaging, so I personally prefer the approach of just telling people about the new way so that they can learn sooner rather than later.

olebole commented:
I would like to put another CI case here, which are integration tests for packages: For the Debian packages, it is very helpful to be able to test the complete package, on its final place, with all the dependencies installed. Our infrastructure then runs this whenever a dependency is changed, which is not limited to the Python dependencies, but also f.e. cfitsio or external programs.
This was a kind-of struggle with the `python setup.py test`, one then has to find out what really happened; but with pytest it is convenient (and standard, which makes my life easier).

mhvk commented:
@astrofrog - OK, point taken about it being better not to use `setup.py test` any more.

I now tried `tox` and indeed it is not much slower than the current `setup.py test` - as you say, both create a temporary environment. But at first, it didn't run the tests in the docs, and after adding a suitable `addopts = --doctest-rst` to `setup.cfg`, it fails all the tests... (somehow I seem to need `pytest-3` rather than `pytest`...). I was also not so pleased with now having 140MB worth of downloaded packages under `.tox`. I guess this echoes @cdeil's comment that I much prefer to just run tests with the standard system packages.

Anyway, a bit more reading of docs and experimenting led me to the following, which does roughly what I would need (separately: I just put it in `setup.cfg`; maybe eventually one would move to `pyproject.toml`, but I'd definitely want things together, so am not keen on `tox.ini`)
```
[testenv:test]
sitepackages = true
commands =
  python -m pytest scintillometry docs

[testenv:build_docs]
sitepackages = true
commands =
  python -m sphinx docs docs/_build/html -W -b html
```

p.s. One pity is that by default the nice pytest header disappears for affiliated packages - looks like a `conftest.py` is needed in the main directory - I'll raise a separate issue for that in `astropy-template`.

astrofrog commented:
For the header, I've noticed that before and I think this might be to actually make the header plugin into a proper plugin rather than require a ``conftest.py`` file to be added for all packages. Then we could opt in to it by having an option in the pytest section of ``setup.cfg`` and maybe even configure there what packages to show (cleaner than relying on ``conftest.py`` files).

astrofrog commented:
For the site-packages option of tox, note that you can also run tox with the ``--sitepackages`` flag (in case we don't want it to be included by default in CI for example).

mhvk commented:
:+1: to the display header plugin. It might then make sense to get the things listed from the dependencies already given. Could it be part of `pytest-astropy`?

And thanks for the `--sitepackages` flag. I guess easiest might be to have it set in the default "test" environment, but not set it in environments actually tried in CI. (I do like the idea that I can (try to) reproduce CI problems by trying `tox -e <CI env>`)

astrofrog commented:
@mhvk - regarding the display plugin - yes it could be part of pytest-astropy and in fact we had an issue open for it (https://github.com/astropy/pytest-astropy/issues/15). I'll work on this now.

mhvk commented:
One more thing I regularly use is `setup.py test --coverage` - would be good to add its replacement as well.

cdeil commented:
@mhvk - What about this?
```
pytest -v scintillometry --cov= scintillometry --cov-report=html
```
This will generate the coverage report as before, no?

Cadair commented:
Yeah and it works with xdist 😉

mhvk commented:
Thanks, indeed, after installing `python3-pytest-cov`, things work (removing the space after `cov=`), though it does look like I should set up a coverage section to exclude the test files themselves.

In any case, something to mention in the APE (and perhaps also add a dependency on `pytest-cov` in `pytest-astropy`?)

cdeil commented:
@mhvk - You can configure the coverage from a `.coverage` file, or put the config in `setup.cfg` (and probably other ways).

If you don't want to measure coverage for your test files, it's easy to omit.
Example of what we put (mostly copy & paste from the Astropy package template): https://github.com/gammapy/gammapy/blob/983d3055e477125bb2449fe1d6c669372a7a53b0/setup.cfg#L94-L118

mhvk commented:
OK, thanks. I can also see how it would be easy to create a `tox -e test-coverage` section.

Sorry, to keep on harassing, but another thing I like about `setup.py` is the built-in help. E.g., `setup.py test --help` tells me that I need to pass on `--coverage`. I guess `tox -a` may help a bit here.

cdeil commented:
@mhvk - You can get help on the options from pytest directly, it also includes the help on options from plugins.
```
$ pytest -h | grep cov
coverage reporting with distributed testing support:
  --cov=[SOURCE]        Path or package name to measure during execution
                        (multi-allowed). Use --cov= to not do any source
  --cov-report=TYPE     Type of report to generate: term, term-missing,
                        missing may be followed by ":skip-covered". annotate,
                        specifies the output location. Use --cov-report= to
  --cov-config=PATH     Config file for coverage. Default: .coveragerc
  --no-cov-on-fail      Do not report coverage if test run fails. Default:
  --no-cov              Disable coverage report completely (useful for
  --cov-fail-under=MIN  Fail if the total coverage is less than MIN.
  --cov-append          Do not delete coverage but append to current. Default:
  --cov-branch          Enable branch coverage.
                        discovery
                        prefixes or glob names for Python test class discovery
                        method discovery
```

For Gammapy what we've done is to add a [Makefile](https://github.com/gammapy/gammapy/blob/master/Makefile) so that all that devs and maintainers have to remember is to type `make` and get a help printout on all the common commands (see [here](https://gist.github.com/cdeil/b712a97cb0f7fed24a79a5d1aa62305d)). Really nothing else, just to type `make`.

I haven't used `tox` yet (except played with it briefly), but I think you could use it in the same way -- pre-define the common commands there, and also add some help printout with further information to make the live of developers easy.

mhvk commented:
Hmm, that `Makefile` is not a bad idea. Wonder if a more python specific standard will arise...

cdeil commented:
@mhvk - Makefiles are nice, but also have big issues. If you use spaces instead of tabs it gives cryptic errors, very weird for Python people. And it doesn't work on Windows.

I think the choice made here in the APE to recommend tox is probably the best option. It's not a standard, but the tagline at https://tox.readthedocs.io is "vision: standardize testing in Python" and it's widely used by many Python projects. There's also https://nox.thea.codes which seems nice if you want to automate your tasks with Python instead of a config file, but I think tox is much more widely used.

@astrofrog - Apologies for derailing the discussion here. Overall I'm strongly 👍 for this APE. There are issues and details that are complex, IMO especially concerning pyproject.toml and where builds take place in the dev workflow. My largest concern for distribution is that you recommend to pin on old Numpy versions for the build env, and then some users will get install errors because the build of that old Numpy will fail on their machine. E.g. you don't know if it'll work on the next version of macOS. But those are things we will see and figure out over the next years, the recommendations given here don't have to be the last word. So 👍  to merge.

Cadair commented:
> because the build of that old Numpy will fail on their machine

I think this should be a seriously edgey edge case, it should be using wheels for this install almost all the time, basically unless you are on a platform which doesn't have wheels built for it (i.e. alpine linux) or similar. When you pin these versions you know if the wheels exist or not, barring someone going back and removing them for old versions, they should stay there.

I agree with you that we can work this out as we go along (which really isn't any different from the equally (more?) complex but more familiar situation we have at the moment).

I think that having a controlled build environment for wheels and sdist fixes more problems than it's likely to introduce on developer's machines, but :man_shrugging:

cdeil commented:
@Cadair - the case of using a modern Python (3.7, 3.8, 4.0) is common, and then you won't have Numpy wheels for older Numpy versions.

Cadair commented:
You should update your pins in `pyproject.toml` to the latest version of numpy for a given Python with a wheel, which I think is in the APE. I appreciate this takes intervention at each Python release.

cdeil commented:
@Cadair - Did you really mean "latest" as most recent? I think @astrofrog was arguing that one should pin against as old versions of Numpy as possible?

When Python 3.8 comes out next month, and only after that release Numpy wheels for 3.8 will appear, what Numpy should one pin against?

Whatever the answer, this is really not a good solution, for each project maintainer to have to watch Python and Numpy releases and who supports which old Numpy version, every time before cutting a new release. I don't want to do that, that's for the packagers (like conda-forge or Debian) to figure out.

Cadair commented:
>  Did you really mean "latest" as most recent?

No I meant oldest for a given Python.

>  that's for the packagers (like conda-forge or Debian) to figure out.

If you are releasing binary wheels of your project you have taken on that role for PyPI packages. The build isolation is designed specifically to handle things that debian and conda are already doing for this reason.

If you aren't building binary wheels then I think it probably matters less what version you pin numpy to (or just don't at all), because in that situation all the build requirements are doing is making sure certain packages are installed before `setup.py` is executed so you can lookup the numpy headers etc.

cdeil commented:
Gammapy is one of the packages that isn't building binary wheels. I think that's the case for most Astropy affiliated packages.

So yes, adding a `pyproject.toml` without Numpy pinned is an option worth considering for us.
- Pro: Solves the installation order issue
- Con: Creates extra virtual envs during install for users

I'm still a bit fuzzy on what exactly would happen. Gammapy depends on `numpy`, `astropy`, `astropy-regions` and soon `astropy-healpix` which all have C extensions. So potentially if no wheels are available for some of those (which is common with new Python versions), the build of each of those would create their own venv and install Numpy several times during `pip install gammapy`, of those packages had `pyproject.toml` files?
Seems pointless if in the end they will all have to work together in the one user env that already exists when starting pip.

mhvk commented:
I'm also happy with the APE - I only would suggest adding a line or two about `--coverage` (maybe both the `tox` setup and the use with `pytest-cov` from @cdeil above) - it is the one thing we do in travis which currently is missing.

For eventual documentation (and perhaps for here), I would also swap the order: describe first how test/coverage/docs are done by the individual commands and only then describe how the current `astropy-helpers` + `setup.py`automation can be easily replaced with `tox`.

Cadair commented:
> which all have C extensions

Just for clarity, you only need to put packages in the `pyproject.toml` file if they are *build* time dependencies. So this is normally only `numpy` if you are building a C extension, or maybe `numpy` and `Cython` if you are build a Cython extension (and using numpy arrays in them, which I am assuming you are lol!).

> the build of each of those would create their own venv and install Numpy several times

When you are installing just from sdists, and you have a chain of dependencies which are linking against numpy, pip would build the numpy wheel once, and then use it for all the following builds. It would use multiple isolated build envs though. (Which is pretty cheap when you are only installing numpy, setuptools, wheel and maybe cython into it).

> Seems pointless if in the end they will all have to work together in the one user env that already exists when starting pip.

It allows each of those packages to use different versions of the build time dependencies though. For instance different Cython versions or different versions of `astropy_helpers` (or whatever a build time dep version of it gets called). Which is one of the things the submodule gives us and this mechanism makes redundant.

The overhead of doing this is really pretty low (maybe if you are unlucky one extra numpy compile time) and the advantages are massive both for wheels and for having control over your build environment, which is something setuptools has no way of doing without PEP 517.

astrofrog commented:
Just for anyone following this - my plan is now to write up some package maintainer-facing packaging docs that assume the infrastructure described here.

@mhvk: I think I'd rather keep the order of things as-is in the APE, and focus instead on writing good docs for maintainers - ideally I don't think maintainers should have to read the APE in detail, as APEs shouldn't be substitutes for documentation (and the order of things in the APE makes more sense to me from the point of view of a proposal for change).

mhvk commented:
@astrofrog - yes, definitely OK to keep the APE as is if you are writing more general documentation!

pllim commented on APE17.rst:
```diff
@@ -0,0 +1,593 @@
+described in its documentation and we therefore do not repeat here. However, one
+subtlety is that since it relies on tags to determine versions, for packages
+such as the core astropy package which have all their tags on branches (at least
+in recent years), we will need to add a 'developer' tag on ``master`` straight
```
@jdavies-st , does this sufficiently address your concern about dev tags?

jdavies-st replied:
Yes, it does.  So that would be something like a `4.1.a0` tag on `master` so that the version `setuptools_scm` reports would be something meaningful, such as `astropy 4.1.a0.dev23` meaning that `master` is now 23 commits after version `4.0.0` was tagged (on its own branch).  That `a0` is important if you plan on using conda for anything with dev versions of astropy, but otherwise is probably not necessary.

Cadair replied:
according to the version numbering PEP `.dev` is also fine for that, you would then get a version like `4.1.dev23`

pllim replied:
`astropy` never plans to have a dev release on `conda` though, so is the `a0` relevant then?

jdavies-st replied:
No.  We were doing nightly dev builds of conda packages for `jwst`, so it was relevant to us.  Not relevant here though. This looks good to me. 👍

Cadair commented on APE17.rst:
```diff
@@ -500,10 +502,14 @@ and their proposed replacement:
+* Determining flags for OpenMP compilation
+
+This package would need to be declared as a build-time dependency in
+``pyproject.toml`` for packags that wish to use it.
```
packages

bwinkel replied:
It would be great, if the `extension-helpers` could also figure out build/linker flags for C++ (if used), because this is sometimes a really pain, especially, if C++11 features are desired and one wants to build on multiple platforms.

astrofrog replied:
Indeed there's actually no reason that extension-helpers couldn't be extended to help with extensions beyond C (e.g. C++, rust, etc.) as long as we keep it general.

eteq commented:
Following some discussions at the Astropy Coordination Meeting (mostly with @astrofrog and myself), I've identified a few things to add to this APE (or possibly say "out of scope for the APE, but should do in the implementation stage"):

1. We should commit to a "migration guide" for users of the existing helpers to switch to the new scheme.  Tentatively it would be generated based on the Astropy core experience, and would be "released" once the process is done in the core.
2. We should state a deadline we're aiming for to end support for the "old" `astropy-helpers`.  2 years (i.e. Astropy LTS 5.0) might be a good deadline.  Dependent on the "migration guide" being done and demonstrated successful (i.e., "used" by some affiliated or coordinated package)
3. Relatedly, we should be sure the package template is up to date and at least as usable (maybe more, with these changes!) before we can declare victory on the transition.

astrofrog commented:
@eteq - I think the following section should now address all your points:

https://github.com/astropy/astropy-APEs/blob/7f2b4dabb9386a47697e0e4a0da5499804e75386/APE17.rst#timeline-for-changes

Please check whenever you can and confirm whether you agree!

taldcroft commented:
The new Timeline for changes section looks fine to me, so approving this.  In the second paragraph there are a few instances of the word "should" that might be changed to "shall" or "will" (e.g. " the migration guide should be advertised widely to all maintainers") to convey a more definitive requirement, but there is no doubt that these things will happen so this change is optional to me.

hamogu commented on APE17.rst:
```diff
@@ -0,0 +1,701 @@
+
+We could then provide a clear documentation page in a centralized place on how
+to use pytest and how to do common things the test command used to allow (e.g.
+running tests for specific sub-packages, or with remote data).
```
I think the parentheses in this sentence should be removed.

astrofrog replied:
Done!

eteq commented on APE17.rst:
```diff
@@ -0,0 +1,701 @@
+<http://docs.astropy.org/en/stable/development/releasing.html>`_ as well as
+`packages using the astropy package template
+<http://docs.astropy.org/en/stable/development/astropy-package-template.html#releasing-a-python-package>`_
+will need to be updated to reflect the use of setuptools_scm.
```
I think the template needs to be updated for quite a bit more than just `setuptools_scm`, right?  That probably should be stated explicitly as a mandatory part of the process.

eteq replied:
I guess it's in the "timeline" section, although it's not clear exactly what makes this special, then?

Cadair replied:
This is under the "Version Helpers" section, and is describing things which reference the old version stuff that would need adjusting for the new version stuff.

astrofrog replied:
Fixed in https://github.com/astropy/astropy-APEs/pull/52/commits/7695c58e8e539acf8dc1f94e741448a1045c4d7e

eteq commented on APE17.rst:
As discussed in the coordination meeting, the one thing it will do is make it so that releases no longer contain c files.  I'm OK with that, but it should be noted here so we're sure we all understand.

Cadair replied:
I don't think this is the right section for that. I wouldn't consider someone who cares if the tarball contains pre-generated or auto-generated C files our regular user. I am also pretty sure we mention it further down the document.

Cadair replied:
This is elaborated upon in [this](https://github.com/astropy/astropy-APEs/pull/52/files#diff-7e96e0f54f0335dc6dfa73709a7c40d7R455) section.

eteq commented on APE17.rst:
I think it's worth stating that `pip` will probably eventually break the helpers (if that's true?).  That's a stronger "status quo won't work"/there are no alternatives argument.

eteq commented on APE17.rst:
Is it worth pointing out that we *can* potentially keep these commands, if they translate into `tox` or `pytest` commands?

Cadair replied:
I added a note to this effect in the "alternatives" section.

eteq commented on APE17.rst:
```suggestion
Impact for users who are not developers
```

because on some level the "non-core developers" are really just users - this makes it less ambiguous

astrofrog replied:
Done

eteq commented:
Thanks for addressing all my last bits there, @Cadair.  And thanks to @astrofrog and @Cadair for all their hard work on this.  

Or as @cadair would say: *burn it all*

eteq commented:
This APE has been accepted following unanimous approval at the 2019 coordination meeting.
